docs/incubator-proposal.md

```diff
@@ -0,0 +1,194 @@
+# cert-manager proposal
+
+## Problem
+
+Currently in Kubernetes, a number of different components including user
+applications require TLS certificates. Right now, users can provide
+certificates to workloads on the cluster by creating `Secret` resources that
+contain a private/public keypair.
+
+A number of projects (see: [kube-lego](https://github.com/jetstack/kube-lego),
+[kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager)
+and [kubernetes-letsencrypt](https://github.com/tazjin/kubernetes-letsencrypt))
+came about in order to automate the retrieval of certificates from an [ACME](https://github.com/ietf-wg-acme/acme/)
+compliant server. This has served the community well, especially with the rise
+in popularity of ACME due to [letsencrypt](https://letsencrypt.org). These
+projects also attempt to renew the certificate as it reaches it's expiry time.
+
+Whilst each project has invented it's own schema/methodology for obtaining
+these certificates, the end goal has largely been the same. To provide a signed
+keypair in a Secret resource to then be consumed, either by an Ingress
+controller or some other workload in-cluster, and then keep that keypair up to
+date.
+
+Looking beyond the capabilities of all of these projects, it'd also be
+beneficial for users to be able to utilise alternative issuer backends, other
+than ACME. This could include [Hashicorp Vault](https://vaultproject.io), a
+plain CA (backed by Kubernetes secret resources), or any other cloud based
+service.
+
+It is therefore proposed that a new project, `cert-manager`, is created that
+aims to generalise the issuance and renewal of certificates from an arbitrary
+certificate source. This project should be designed in a similar manner to
+other Kubernetes controllers (eg. kube-controller-manager, service-catalog et
+al.).
```

Comment:
While I agree this is a useful goal I think the scope may be larger than anticipated. Interacting with most traditional certificate authorities requires the implementation of a state machine that models the authorize, submit, wait, retreive cycle (and variations on that theme).

Experience with the OpenStack barbican project has shown that this quickly becomes a non-trivial problem to solve.

Reply:
Authorization to the CA is a huge problem here, understanding how that's done, where those secrets would be managed becomes very ugly.

Reply:
Which goal in particular do you think doesn't fit? Designing it in a control-loop like fashion?

I think your point here about traditional authorities is valid, but how this is achieved can be implemented in a per-CA way. We don't generalise the mechanism for storing secrets for each issuer, that is issuer specific and (especially in the secrets case) aims to follows K8s best practices (secrets are stored in Secrets).

Each Issuer has it's own `status` block on a resource, allowing it to store information about a request and/or references to external resources (e.g. Secrets) containing values required during auth, submit, wait or retrieval. An issuer can either do this synchronously (as is *currently* the case with the ACME implementation) or asynchronously, by storing state back in the apiserver (either in the Status block, or resources referenced by the Certificate/Issuer resource in some way).

Reply:
I'm not saying anything here is wrong, just that the async CA's get a bit hard to handle over time. The experience with deployers of OpenStack was that only the CA integrations that got written and provided by default got used, I'm not aware of any that got written and contributed upstream by CAs or consumers.

Reply:
Most enterprises tend to have one or two "corporate" CA's - they want all the applications running on their cloud to be signed with certificates from that CA but rarely if ever are prepared to provide credentials that allow certificates to be retrieved in an automated way.

It's perhaps looking a bit more closely at what you expect the use cases to be. I worked on a project a few years ago (at a previous company) where one of our cloud orchestration products built integrations for various CA's - so that you could deploy a cloud on-prem and it would pretty much have TLS everywhere but more importantly had the capability to either get all certificates from the corporate CA (with appropriate credentials) or sign with a delegated signing authority; in both cases we found that the overwhelming majority of companies were not happy with providing credentials for blanket access to the CA (which I agree with) or delegated signing authorities (which I agree with a bit less).

Reply:
👍 ack. To be honest, by flexible I don't mean necessarily super-easy to implement new CAs. I think no matter what we're going to have to make PRs for this kind of support. I think we also wouldn't want anything like out of tree CA implementations as I think it'd be difficult to control quality.

As we also don't *require* that the flow is asynchronous, I'd hope we have a fair amount of wiggle room when implementing future CAs. Time will tell though...

Reply:
Sorry I've just seen your most recent follow up.

So I'd imagine that a cluster admin would create the 'Issuer' resources for a cluster. So in the described enterprise case, an admin (someone who *does* have authorisation to 'touch' CA credentials) would create the appropriate Issuer for the environment. This access control can be enforced through RBAC, and users can be disallowed for reading/writing Issuer resources.

Users are then able to create `Certificate` resources (which are more akin to CSRs than certificates themselves), which specify which Issuer to obtain the certificate from. cert-manager is then responsible for authorising the request (through negotiation with the CA), and obtaining the certificate. This allows for the separation between those that have permission to access the CA, and those with permission to obtain certificates.

Things that need exploring:

* We don't support non-namespaced Issuers right now, meaning an issuer needs to be creating in each namespace. This kind of goes against the cluster-admin pattern, as it means that a user needs to restricted from accessing certain Secret and Issuer resources in their own namespace, which quickly becomes hard to manage. We've got #91 to track this.

* There's currently no way for an Issuer resource to restrict what domains it can issue - I'd imagine a lot of these organisations want to restrict who can request what certificates. This could be implemented at the CA level itself, or it could be implement in cert-manager. Whilst my preference is towards it being the CAs responsibility (and so the CA could 'deny' the authorisation during the auth stage), I am certain some users would prefer to specify this at the cert-manager level.

* We currently don't have the concept of 'requester identity'. Right now there's no way to identify who created a certificate request aside from looking at the kube-apiserver audit log. The kubelet TLS bootstrapping API does have fields to track this (see: https://github.com/kubernetes/kubernetes/pull/25562), and it's something we should definitely consider. This may become especially important with a vault issuer, depending on how we implement identity there.